### PR TITLE
Support simple assertions

### PR DESCRIPTION
**Pull request by jayvdb**
Also include test cases for the harder assertions.

Related to https://github.com/google/sre_yield/issues/6

### CODE REVIEW

**Comment by jayvdb** [comment]
This is only solving the most basic cases, but I happen to have a dataset with a whole lot of them, and that is the main problem I am encountering.  I can easily pre-process my regex to solve it without the PR, and I fully understand if you would rather have no solution instead of this.

It seems solving this 'properly', even partially, would require adding more states, and a lot more complexity.  I'm not sure I am up to attempt that atm.  What I could see being useful is for the current PR to also raise an exception for the cases that cant be handled - I think most of them can be detected without altering the structure/design of the current project.  I would be happy to attempt that.

File: sre_yield/__init__.py [hunk 1]
```diff
@@ -563,8 +563,8 @@ def __init__(self, pattern, flags=0, charset=CHARSET, max_count=None):
             sre_constants.MIN_REPEAT: self.max_repeat_values,
             sre_constants.MAX_REPEAT: self.max_repeat_values,
             sre_constants.AT: self.nothing_added,
-            sre_constants.ASSERT: self.empty_list,
-            sre_constants.ASSERT_NOT: self.empty_list,
+            sre_constants.ASSERT: self.maybe_save,
+            sre_constants.ASSERT_NOT: self.nothing_added,
```
Comment (thatch) [inline sre_yield/__init__.py]: If you make `__init__` take some a `relaxed` flag, setting these both to `nothing_added` is good.  You can probably skip `check_anchor_state` when in that mode as well.

File: sre_yield/tests/test_sre_yield.py [hunk 2]
```diff
@@ -58,6 +58,36 @@ def testAlternationWithEmptyElement(self):
         self.assertSequenceEqual(sre_yield.AllStrings("a[bc]?"), ["a", "ab", "ac"])
         self.assertSequenceEqual(sre_yield.AllStrings("a[bc]??"), ["a", "ab", "ac"])
 
+    def testHttp(self):
+        self.assertSequenceEqual(sre_yield.AllStrings("^http://"), ["http://"])
+
+    def testLookaroundNegativeNop(self):
+        self.assertSequenceEqual(sre_yield.AllStrings("(?!b)c"), ["c"])
+        self.assertSequenceEqual(sre_yield.AllStrings("(?!c)c"), ["c"])
```
Comment (thatch) [inline sre_yield/tests/test_sre_yield.py]: This one is currently a bug, should be empty list.

File: sre_yield/tests/test_sre_yield.py [hunk 3]
```diff
@@ -58,6 +58,36 @@ def testAlternationWithEmptyElement(self):
...
+        self.assertSequenceEqual(sre_yield.AllStrings("(?!b)c?"), ["", "c"])
+        self.assertSequenceEqual(sre_yield.AllStrings("a(?!b)c"), ["ac"])
+        self.assertSequenceEqual(sre_yield.AllStrings("a(?!b)c?"), ["a", "ac"])
+        self.assertSequenceEqual(sre_yield.AllStrings("a(?!b)$"), ["a"])
+
+        self.assertSequenceEqual(sre_yield.AllStrings("(?<!b)c"), ["c"])
+        self.assertSequenceEqual(sre_yield.AllStrings("^(?<!b)c"), ["c"])
+        self.assertSequenceEqual(sre_yield.AllStrings("a(?<!b)c"), ["ac"])
```
Comment (thatch) [inline sre_yield/tests/test_sre_yield.py]: I think `b(?<!b)c` is also a bug, if you're codifying current behavior.

File: sre_yield/tests/test_sre_yield.py [hunk 4]
```diff
@@ -58,6 +58,36 @@ def testAlternationWithEmptyElement(self):
...
+
+    def testLookaroundPositiveAdd(self):
+        self.assertSequenceEqual(sre_yield.AllStrings("(?<=abc)def"), ["abcdef"])
```
Comment (thatch) [inline sre_yield/tests/test_sre_yield.py]: This one doesn't make a sense to me if we're treating the matches as fullmatch only; the `abc` part would be before the beginning of the match.  Is there an equivalent one like `xxx(?<=abc)def` that demonstrates what you want to show as well?

File: sre_yield/__init__.py (hunk 1 above)
Comment (jayvdb) [reply sre_yield/__init__.py]: Ok, a `relax` flag sounds like a good compromise.

Comment (jayvdb) [reply sre_yield/__init__.py]: this PR now redundant as https://github.com/google/sre_yield/pull/21 does the trick.

**Pull request closed without merging**